### PR TITLE
Attribute/Annotations (WIP)

### PR DESCRIPTION
**Pull request by retailcoder**
This is still work-in-progress, but important changes have been made to parse tree inspections in order to allow them to work off the attributes pass' lexer tokens.

The inspections are implemented and yield results, but the *synchronize attributes/annotations* quickfix isn't completed. I'm pulling the changes in anyway, since I don't know when I'll find the time to implement that quickfix and I don't want that to delay the structural changes in here.

### CODE REVIEW

File: Rubberduck.Parsing/VBA/ComponentParseTask.cs [hunk 1]
```diff
@@ -48,12 +51,18 @@ public void Start(CancellationToken token)
...
 
                 var tokenStream = RewriteAndPreprocess(token);
                 token.ThrowIfCancellationRequested();
 
-                var attributes = _attributeParser.Parse(_component, token);
+                Logger.Trace($"ParseTaskID {_taskId} begins attributes pass.");
+                ITokenStream attributesTokenStream;
+                IParseTree attributesTree;
+                var attributes = _attributeParser.Parse(_component, token, out attributesTokenStream, out attributesTree);
+                Logger.Trace($"ParseTaskID {_taskId} finished attributes pass.");
```
Comment (MDoerner) [inline Rubberduck.Parsing/VBA/ComponentParseTask.cs]: This should get extracted unto a separate method.

File: Rubberduck.Parsing/VBA/DeclarationResolveRunnerBase.cs [hunk 2]
```diff
@@ -49,8 +49,6 @@ public abstract class DeclarationResolveRunnerBase : IDeclarationResolveRunner
 
         protected void ResolveDeclarations(QualifiedModuleName module, IParseTree tree, CancellationToken token)
         {
-            if (module == null) { return; }
```
Comment (MDoerner) [inline Rubberduck.Parsing/VBA/DeclarationResolveRunnerBase.cs]: Why has this been removed? Now, passing in null results in a NRE in the catch block.

File: Rubberduck.Parsing/VBA/ParseRunnerBase.cs [hunk 3]
```diff
@@ -5,23 +5,25 @@
...
+        private readonly IModuleExporter _exporter;
 
-
-        public ParseRunnerBase(
+        protected ParseRunnerBase(
             RubberduckParserState state,
             IParserStateManager parserStateManager,
             Func<IVBAPreprocessor> preprocessorFactory,
-            IAttributeParser attributeParser)
+            IAttributeParser attributeParser, IModuleExporter exporter)
```
Comment (MDoerner) [inline Rubberduck.Parsing/VBA/ParseRunnerBase.cs]: That should be two lines.

File: Rubberduck.Parsing/VBA/RubberduckParserState.cs [hunk 4]
```diff
@@ -326,7 +328,7 @@ private void OnStateChanged(object requestor, ParserState state = ParserState.Pe
             HandleStateChanged(state);
 
             Logger.Info($"{nameof(RubberduckParserState)} ({_stateChangedInvocations}) is invoking {nameof(StateChanged)} ({Status})");
-            StateChanged?.Invoke(requestor, new ParserStateEventArgs(state));
+             StateChanged?.Invoke(requestor, new ParserStateEventArgs(state));
```
Comment (MDoerner) [inline Rubberduck.Parsing/VBA/RubberduckParserState.cs]: Superfluous whitespace

File: Rubberduck.Parsing/VBA/DeclarationResolveRunnerBase.cs (hunk 2 above)
Comment (retailcoder) [inline Rubberduck.Parsing/VBA/DeclarationResolveRunnerBase.cs]: `QualifiedModuleName` is a struct... it *can't* be `null` ;-)

Comment (MDoerner) [inline Rubberduck.Parsing/VBA/DeclarationResolveRunnerBase.cs]: Ah, that explains it.

File: RetailCoder.VBE/Common/ModuleExporter.cs [hunk 5]
```diff
@@ -10,17 +10,11 @@ public class ModuleExporter : IModuleExporter
...
-                    ? ApplicationConstants.RUBBERDUCK_TEMP_PATH
-                    : Path.GetDirectoryName(Assembly.GetAssembly(typeof(App)).Location);
-            }
-        }
+        public string ExportPath => TempFile
+            ? ApplicationConstants.RUBBERDUCK_TEMP_PATH
+            : Path.GetDirectoryName(Assembly.GetAssembly(typeof(App)).Location);
 
-        public string Export(IVBComponent component, bool tempFile = false)
+        public string Export(IVBComponent component, bool tempFile = true)
```
Comment (Vogel612) [inline RetailCoder.VBE/Common/ModuleExporter.cs]: this is a pretty breaking change... are we sure this doesn't affect any callers incorrectly?

File: RetailCoder.VBE/Settings/CodeInspectionSettings.cs [hunk 6]
```diff
@@ -137,7 +135,11 @@ public CodeInspectionSetting()
             //default constructor required for serialization
         }
 
-        public CodeInspectionSetting(string name, string description, CodeInspectionType type, CodeInspectionSeverity defaultSeverity, CodeInspectionSeverity severity)
+        public CodeInspectionSetting(string name, CodeInspectionType type, CodeInspectionSeverity defaultSeverity = CodeInspectionSeverity.Warning)
+            : this(name, string.Empty, type, defaultSeverity, defaultSeverity)
+        { }
+
+        public CodeInspectionSetting(string name, string description, CodeInspectionType type, CodeInspectionSeverity defaultSeverity = CodeInspectionSeverity.Warning, CodeInspectionSeverity severity = CodeInspectionSeverity.Warning)
```
Comment (Vogel612) [inline RetailCoder.VBE/Settings/CodeInspectionSettings.cs]: that's a long line there ... can we make that less scroll-requiring?

File: RetailCoder.VBE/UI/CodeExplorer/Commands/AddClassModuleCommand.cs [hunk 7]
```diff
@@ -15,12 +15,12 @@ public AddClassModuleCommand(AddComponentCommand addComponentCommand) : base(Log
...
         }
 
-        protected override bool CanExecuteImpl(object parameter)
+        protected override bool EvaluateCanExecute(object parameter)
         {
             return _addComponentCommand.CanAddComponent(parameter as CodeExplorerItemViewModel);
         }
 
-        protected override void ExecuteImpl(object parameter)
+        protected override void OnExecute(object parameter)
```
Comment (Vogel612) [inline RetailCoder.VBE/UI/CodeExplorer/Commands/AddClassModuleCommand.cs]: suggests an event .. is it though? `DoExecute` or even `Execute` seems like the better name to me...

File: RetailCoder.VBE/UI/Command/CommandBase.cs [hunk 8]
```diff
@@ -15,24 +15,18 @@ public abstract class CommandBase : ICommand
 
         protected CommandBase(ILogger logger)
         {
-            _logger = logger;
+            Logger = logger;
```
Comment (Vogel612) [inline RetailCoder.VBE/UI/Command/CommandBase.cs]: doesn't this require `Logger` to have a `private set;`? or possibly a `readonly`?

File: RetailCoder.VBE/UI/RubberduckUI.resx [hunk 9]
```diff
@@ -2082,4 +2082,10 @@ Would you like to import them to Rubberduck?</value>
   <data name="RenameDialog_PropertyRenameError" xml:space="preserve">
     <value>Could not rename Property {0}.</value>
   </data>
+  <data name="CodeInspectionSettings_Performance" xml:space="preserve">
+    <value>Performance</value>
+  </data>
+  <data name="CodeInspectionSettings_RubberduckOpportunities" xml:space="preserve">
+    <value>Rubberduck Opportunities</value>
+  </data>
```
Comment (Vogel612) [inline RetailCoder.VBE/UI/RubberduckUI.resx]: Note to i18n: needs to be translated for next release

File: Rubberduck.Inspections/Concrete/UnassignedVariableUsageInspection.cs [hunk 10]
```diff
@@ -11,6 +12,7 @@
 
 namespace Rubberduck.Inspections.Concrete
 {
+    [SuppressMessage("ReSharper", "LoopCanBeConvertedToQuery")]
```
Comment (Vogel612) [inline Rubberduck.Inspections/Concrete/UnassignedVariableUsageInspection.cs]: I like this a lot more than the magic comments for `ReSharper DisableOnce: ...`

File: Rubberduck.Inspections/Inspector.cs [hunk 11]
```diff
@@ -60,25 +61,52 @@ private void UpdateInspectionSeverity(Configuration config)
...
+                }
+                
                 // Prepare ParseTreeWalker based inspections
-                WalkTrees(config.UserSettings.CodeInspectionSettings, state, _inspections.OfType<IParseTreeInspection>());
+                var passes = Enum.GetValues(typeof (ParsePass)).Cast<ParsePass>();
+                foreach (var parsePass in passes)
+                {
+                    try
+                    {
+                        WalkTrees(config.UserSettings.CodeInspectionSettings, state, parseTreeInspections.Where(i => i.Pass == parsePass), parsePass);
```
Comment (Vogel612) [inline Rubberduck.Inspections/Inspector.cs]: I wonder why the relevant inspections aren't bundled into the pass, but instead queried from all inspections like this.. It seems like a weird way to organize things

File: Rubberduck.Inspections/Inspector.cs [hunk 12]
```diff
@@ -103,18 +131,34 @@ private void UpdateInspectionSeverity(Configuration config)
...
-                    inspections.Where(i => i.Severity != CodeInspectionSeverity.DoNotShow && !IsDisabled(settings, i))
-                        .Select(inspection =>
-                        {
-                            inspection.Listener.ClearContexts();
-                            return inspection.Listener;
-                        })
-                        .ToList();
+                var listeners = inspections
+                    .Where(i => i.Severity != CodeInspectionSeverity.DoNotShow
+                        && i.Pass == pass
```
Comment (Vogel612) [inline Rubberduck.Inspections/Inspector.cs]: why do inspections know which pass they belong to? this seems weird (again)

File: Rubberduck.Inspections/ParseTreeListeners/AttributeAnnotationListener.cs [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+            CurrentScopeDeclaration = HasMembers ? Members.Value[memberName] : Module.Value;
+        }
+
+        public override void EnterModule(VBAParser.ModuleContext context)
+        {
+            Module = new Lazy<Declaration>(() => State.DeclarationFinder
+                .UserDeclarations(DeclarationType.Module)
+                .SingleOrDefault(m => m.QualifiedName.QualifiedModuleName.Equals(CurrentModuleName)));
+
+            Members = new Lazy<IDictionary<string, Declaration>>(() => State.DeclarationFinder
```
Comment (Vogel612) [inline Rubberduck.Inspections/ParseTreeListeners/AttributeAnnotationListener.cs]: I'm not sure the Lazy gives us real speed advantages here and it seems like more work ... has anyone benchmarked this against non-lazy Module and Members?

File: Rubberduck.Inspections/QuickFixProvider.cs [hunk 14]
```diff
@@ -170,7 +170,19 @@ public void FixAll(IQuickFix fix, Type inspectionType, IEnumerable<IInspectionRe
                 var modules = filteredResults.Select(s => s.QualifiedSelection.QualifiedName).Distinct();
                 foreach (var module in modules)
                 {
-                    _state.GetRewriter(module).Rewrite();
+                    var moduleRewriter = _state.GetRewriter(module);
+                    if (moduleRewriter.IsDirty)
```
Comment (Vogel612) [inline Rubberduck.Inspections/QuickFixProvider.cs]: this seems like a weird name for a flag indicating that the module rewriter wants to perform actions. I'd expect `IsDirty` to force a refresh instead... maybe `HasActions` or something along these lines could be less of a tripwire

File: Rubberduck.Parsing/Annotations/AnnotationType.cs [hunk 15]
```diff
@@ -1,18 +1,65 @@
-﻿namespace Rubberduck.Parsing.Annotations
+﻿using System;
+
+namespace Rubberduck.Parsing.Annotations
 {
+    /// <summary>
+    /// Member names are 
+    /// </summary>
     public enum AnnotationType
```
Comment (Vogel612) [inline Rubberduck.Parsing/Annotations/AnnotationType.cs]: Shouldn't this be given `[Flags]`?

File: RetailCoder.VBE/UI/CodeExplorer/Commands/AddClassModuleCommand.cs (hunk 7 above)
Comment (retailcoder) [inline RetailCoder.VBE/UI/CodeExplorer/Commands/AddClassModuleCommand.cs]: Rather suggest an event than suggest Java ;-)

File: Rubberduck.Parsing/Annotations/AnnotationType.cs (hunk 15 above)
Comment (retailcoder) [inline Rubberduck.Parsing/Annotations/AnnotationType.cs]: Ugh. And the rest of the summary comment, too!

File: Rubberduck.Inspections/Inspector.cs (hunk 12 above)
Comment (retailcoder) [inline Rubberduck.Inspections/Inspector.cs]: It's not the inspection that needs to know; it's the inspector, needs a way to determine which parse tree to inspect with which inspections. The Pass is just a property of an inspection then; most work off the CodePane pass and aren't affected by this in any way.

File: Rubberduck.Inspections/Inspector.cs (hunk 11 above)
Comment (retailcoder) [inline Rubberduck.Inspections/Inspector.cs]: Not sure what you mean here.

File: RetailCoder.VBE/UI/CodeExplorer/Commands/AddClassModuleCommand.cs (hunk 7 above)
Comment (retailcoder) [inline RetailCoder.VBE/UI/CodeExplorer/Commands/AddClassModuleCommand.cs]: It's not much different than Entity Framework's `OnModelCreating` override.

Comment (retailcoder) [inline RetailCoder.VBE/UI/CodeExplorer/Commands/AddClassModuleCommand.cs]: ("Execute" is already taken)

**Comment by retailcoder** [comment]
ref. #3021: this PR adds a bunch of missing inspections and alpha-sorts the fugly list. At least now it's easier to spot a missing inspection; they show up in the same order as in the *solution explorer*.

**Pull request merged**